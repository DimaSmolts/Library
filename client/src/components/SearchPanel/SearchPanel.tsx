import React from 'react'
import './index.css'

interface SearchPanelProps {
  query: string,
  handleQueryChange: (event: React.FormEvent<EventTarget>) => void,
  handleQuerySubmit: (event: React.FormEvent<EventTarget>) => void,
}

const SearchPanel: React.SFC<SearchPanelProps> = (props: SearchPanelProps) => {
  return <div className='search-panel' >
    <h2>Find a book!</h2>
    <form className='search-panel-form' onSubmit={props.handleQuerySubmit.bind(this)}>
      <span>I am looking for:</span>
      <input className='search-panel-input'
        spellCheck='false'
        type='text'
        placeholder='search...'
        value={props.query}
        onChange={props.handleQueryChange.bind(this)} />
      <input className='search-panel-submit'
        type='submit'
        value='Go!' />
    </form>
  </div>
}

export default SearchPanel
