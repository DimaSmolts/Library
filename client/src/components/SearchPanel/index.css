.search-panel {
  /*
  /* TODO: add fixed height
      padding is not counted (+ 2*20px)
   */
  margin: 1em 0;
  padding: 1em;

  background: var(--dark-blue);
  color: var(--gold);
  border-radius: 20px;
}

.search-panel-form {
  display: grid;
  grid-template-columns: auto 1fr auto;
  grid-gap: 1em;
}

.search-panel-input {
  width: 100%;
  color: var(--gold);
  font-size: 1em;
  border-style: none;
  background: transparent;
  outline: none;
  border-bottom: 1px solid var(--gold);
}

.search-panel-submit {
  font-size: 1em;
  height: 30px;
  width: 100px;
  background: var(--gold);
  color: var(--dark-blue);
  border: 0;
}

.search-panel-submit:hover {
  background: var(--dark-gold);
  cursor: pointer;
}
