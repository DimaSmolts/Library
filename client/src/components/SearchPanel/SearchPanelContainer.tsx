import React from 'react';
import SearchPanel from './SearchPanel';
import { searchBookThunk } from '../../store/thunks/searchThunks';
import { connect, ConnectedProps } from 'react-redux';

interface SearchPanelState {
  query: string,
}

class SearchPanelContainer extends React.Component<PropsFromRedux, SearchPanelState> {
  constructor(props: any) {
    super(props);
    this.state = { query: '' };
    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleQuerySubmit = this.handleQuerySubmit.bind(this);
  }

  handleQueryChange(event: React.FormEvent<EventTarget>) {
    this.setState({ query: (event.target as HTMLInputElement).value });
  }

  handleQuerySubmit(event: React.FormEvent<EventTarget>) {
    event.preventDefault();
    this.props.searchBookThunk(this.state.query);
  }

  render() {
    return <SearchPanel query={this.state.query}
      handleQueryChange={this.handleQueryChange}
      handleQuerySubmit={this.handleQuerySubmit} />
  }
}

const mapDispatch = {
  searchBookThunk: (query: string) => searchBookThunk(query),
}

const connector = connect(null, mapDispatch);
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(SearchPanelContainer);
