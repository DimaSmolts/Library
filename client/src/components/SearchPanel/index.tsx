import React from 'react'
import SearchPanel from './SearchPanel'
import { getBooksThunk, cleanUpThunk } from '../../store/thunks/bookListThunks'
import { connect, ConnectedProps } from 'react-redux'
import { History } from 'history'

interface SearchPanelState {
  query: string,
}

export interface SearchPanelProps {
  queryParams: string,
  history: History,
}

export class SearchPanelContainer extends React.Component<PropsFromRedux & SearchPanelProps, SearchPanelState> {
  constructor(props: PropsFromRedux & SearchPanelProps) {
    super(props)
    const searchQuery = new URLSearchParams(this.props.queryParams).get('search') || ''
    this.state = { query: searchQuery }
    this.handleQueryChange = this.handleQueryChange.bind(this)
    this.handleQuerySubmit = this.handleQuerySubmit.bind(this)
  }

  componentDidMount() {
    this.props.refreshBooks(this.state.query)
  }

  componentWillUnmount() {
    this.props.cleanUpBooks()
  }

  handleQueryChange(event: React.FormEvent<HTMLInputElement>) {
    this.setState({ query: event.currentTarget.value })
  }

  handleQuerySubmit(event: React.FormEvent<HTMLInputElement>) {
    event.preventDefault()
    this.props.refreshBooks(this.state.query)
    this.props.history.push({
      pathname: '/books',
      search: `?search=${this.state.query}`,
    })
  }

  render() {
    return <SearchPanel query={this.state.query}
      handleQueryChange={this.handleQueryChange}
      handleQuerySubmit={this.handleQuerySubmit} />
  }
}

const mapDispatch = {
  refreshBooks: (query: string) => getBooksThunk(query),
  cleanUpBooks: () => cleanUpThunk(),
}

const connector = connect(null, mapDispatch)
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(SearchPanelContainer)
