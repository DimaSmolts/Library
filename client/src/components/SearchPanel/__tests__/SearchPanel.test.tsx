import React from 'react'
import { shallow, mount } from 'enzyme'

import SearchPanel from '../SearchPanel'

const query = 'anything'
const handleChange = jest.fn()
const handleSubmit = jest.fn()

test('snapshot: <SearchPanel /> renders correctly', () => {
  const tree = shallow(
    <SearchPanel
      query={query}
      handleQueryChange={handleChange}
      handleQuerySubmit={handleSubmit} />)

  expect(tree).toMatchSnapshot()
})

it('should submit form', () => {
  const wrapper = mount(<SearchPanel
    query={query}
    handleQueryChange={handleChange}
    handleQuerySubmit={handleSubmit} />)

  wrapper.find('form').simulate('submit')
  expect(handleSubmit).toBeCalled()
})
