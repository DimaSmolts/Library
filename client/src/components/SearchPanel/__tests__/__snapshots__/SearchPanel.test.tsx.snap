// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: <SearchPanel /> renders correctly 1`] = `
<div
  className="search-panel"
>
  <h2>
    Find a book!
  </h2>
  <form
    className="search-panel-form"
    onSubmit={[Function]}
  >
    <span>
      I am looking for:
    </span>
    <input
      className="search-panel-input"
      onChange={[Function]}
      placeholder="search..."
      spellCheck="false"
      type="text"
      value="anything"
    />
    <input
      className="search-panel-submit"
      type="submit"
      value="Go!"
    />
  </form>
</div>
`;
